### ISSUE TITLE
typesystem.xml - Nesting types by specifying '::' is no longer supported (qt_gui_cpp::CompositePluginProvider).

### ISSUE THREAD

**Issue by kjcole** [ros-melodic-arch/ros-melodic-qt-gui-cpp#12]
TL;RL

This is my first go at instaling `ros-melodic-full-desktop`. There's no prior ROS packages on the system. After hours of watching it compile, it died with:

```
qt.shiboken: (typesystem) Error: ~/.cache/yay/ros-melodic-qt-gui-cpp/src/qt_gui_core-0.3.14/qt_gui_cpp/src/qt_gui_cpp_shiboken/typesystem.xml:7:59: Nesting types by specifying '::' is no longer supported (qt_gui_cpp::CompositePluginProvider).
```
Of possible relevance, I have:

- python-shiboken 1.2.4-7
- python-shiboken2 5.14.0-3
- python2-shiboken 1.2.4-7
- shiboken 1.2.4-7
- shiboken2 5.14.0-3
- pyside2 5.14.0-1
- pyside2-tools 5.14.0-1

The "full skinny" console output:

```
==> Making package: ros-melodic-qt-gui-cpp 0.3.14-1 (Mon 13 Jan 2020 05:07:06 AM EST)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found ros-melodic-qt-gui-cpp-0.3.14.tar.gz
==> Validating source files with sha256sums...
    ros-melodic-qt-gui-cpp-0.3.14.tar.gz ... Passed
==> Removing existing $srcdir/ directory...
==> Extracting sources...
  -> Extracting ros-melodic-qt-gui-cpp-0.3.14.tar.gz with bsdtar
==> Sources are ready.
==> Making package: ros-melodic-qt-gui-cpp 0.3.14-1 (Mon 13 Jan 2020 05:07:08 AM EST)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> WARNING: Using existing $srcdir/ tree
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using CATKIN_DEVEL_PREFIX: ~/.cache/yay/ros-melodic-qt-gui-cpp/src/build/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/melodic
-- This workspace overlays: /opt/ros/melodic
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.8.1", minimum required is "2") 
-- Using PYTHON_EXECUTABLE: /usr/bin/python3
-- Using default Python package layout
-- Found PY_em: /usr/lib/python3.8/site-packages/em.py  
-- Using empy: /usr/lib/python3.8/site-packages/em.py
-- Using CATKIN_ENABLE_TESTING: ON
-- Skip enable_testing() when building binary package
-- Using CATKIN_TEST_RESULTS_DIR: ~/.cache/yay/ros-melodic-qt-gui-cpp/src/build/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-3.8
-- catkin 0.7.19
-- BUILD_SHARED_LIBS is on
-- Found Boost: /usr/lib64/cmake/Boost-1.71.0/BoostConfig.cmake (found version "1.71.0") found components: filesystem system 
-- Found TinyXML: /usr/lib/libtinyxml.so  
-- Found PythonLibs: /usr/lib/libpython3.7m.so (Required is at least version "3.8") 
-- Shiboken2Config: Using default python: .cpython-38-x86_64-linux-gnu
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.8.1", minimum required is "3") 
-- Found PythonLibs: /usr/lib/libpython3.7m.so (Required is at least version "3") 
-- SHIBOKEN_PYTHON_INCLUDE_DIRS computed to value: '/usr/include/python3.7m'
-- SHIBOKEN_PYTHON_LIBRARIES computed to value: ''
-- libshiboken built for Release
-- Found PythonLibs: /usr/lib/libpython3.7m.so (Required is at least version "3.8") 
-- Shiboken binding generator available.
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.8.1", minimum required is "3.8") 
-- SIP binding generator available at: /usr/bin/sip
-- Python binding generators: shiboken;sip
-- Configuring done
-- Generating done
-- Build files have been written to: ~/.cache/yay/ros-melodic-qt-gui-cpp/src/build
[  2%] Generating __/__/include/qt_gui_cpp/moc_plugin_context.cpp
[  5%] Generating __/__/include/qt_gui_cpp/moc_plugin.cpp
[  8%] Generating __/__/include/qt_gui_cpp/moc_plugin_bridge.cpp
Scanning dependencies of target qt_gui_cpp
[ 11%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/composite_plugin_provider.cpp.o
[ 14%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/generic_proxy.cpp.o
[ 17%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/plugin_bridge.cpp.o
[ 20%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/plugin_context.cpp.o
[ 23%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/plugin_descriptor.cpp.o
[ 26%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/plugin_provider.cpp.o
[ 29%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/recursive_plugin_provider.cpp.o
[ 32%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/ros_pluginlib_plugin_provider.cpp.o
[ 35%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/settings.cpp.o
[ 38%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/__/__/include/qt_gui_cpp/moc_plugin.cpp.o
[ 41%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/__/__/include/qt_gui_cpp/moc_plugin_bridge.cpp.o
[ 44%] Building CXX object src/qt_gui_cpp/CMakeFiles/qt_gui_cpp.dir/__/__/include/qt_gui_cpp/moc_plugin_context.cpp.o
[ 47%] Linking CXX shared library ../../devel/lib/libqt_gui_cpp.so
[ 47%] Built target qt_gui_cpp
[ 50%] Running Shiboken generator for libqt_gui_cpp Python bindings...
qt.shiboken: (typesystem) Error: ~/.cache/yay/ros-melodic-qt-gui-cpp/src/qt_gui_core-0.3.14/qt_gui_cpp/src/qt_gui_cpp_shiboken/typesystem.xml:7:59: Nesting types by specifying '::' is no longer supported (qt_gui_cpp::CompositePluginProvider).
Cannot parse file: typesystem.xmlshiboken: Error running ApiExtractor.
Command line: --generatorSet=shiboken --include-paths=~/.cache/yay/ros-melodic-qt-gui-cpp/src/qt_gui_core-0.3.14/qt_gui_cpp/src/qt_gui_cpp_shiboken/../../include:/usr/include/qt/:/usr/include/qt/QtWidgets:/usr/include/qt/QtGui:/usr/include/qt/QtCore:/usr/lib/qt//mkspecs/linux-g++ --typesystem-paths=/usr/share/PySide2/typesystems --output-directory=~/.cache/yay/ros-melodic-qt-gui-cpp/src/build/src/qt_gui_cpp_shiboken global.h typesystem.xml
make[2]: *** [src/qt_gui_cpp_shiboken/CMakeFiles/qt_gui_cpp_shiboken.dir/build.make:75: src/qt_gui_cpp_shiboken/libqt_gui_cpp_shiboken/libqt_gui_cpp_shiboken_module_wrapper.cpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:348: src/qt_gui_cpp_shiboken/CMakeFiles/qt_gui_cpp_shiboken.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
==> ERROR: A failure occurred in build().
    Aborting...
checking dependencies...
```

**Comment by jwhendy** [comment]
Hmmm. These lines bother me:
```
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.8.1", minimum required is "2") 
-- SHIBOKEN_PYTHON_INCLUDE_DIRS computed to value: '/usr/include/python3.7m'
-- SHIBOKEN_PYTHON_LIBRARIES computed to value: ''
-- libshiboken built for Release
-- Found PythonLibs: /usr/lib/libpython3.7m.so (Required is at least version "3.8") 
```

This might have nothing to do with it, but could you try building this package after updating the `PKGBUILD` to matc [this one](https://github.com/ros-melodic-arch/ros-melodic-rospack/blob/master/PKGBUILD)? Just these lines:

```
                -DPYTHON_INCLUDE_DIR=/usr/include/python3.8 \
		-DPYTHON_LIBRARY=/usr/lib/libpython3.8.so \
		-DPYTHON_BASENAME=.cpython-38 \
```

The finding of `python3.8` but using also mentioning `python3.7` libs caught my eye. @acxz found that in most packages these variables were added by a script and don't even matter, but generally the build will say those variables are ignored. In yours they are resolving so it's worth a shot.

Hmm. That all said, just googling the error leads me to think you just need to uninstall `shiboken`. Did you google the error? I get relevant hits:

- https://answers.ros.org/question/99351/qt_gui_cpp-fails-to-build-on-debian-jessie/
- that links here: https://github.com/ros-visualization/qt_gui_core/issues/35#issuecomment-28527836
- says the same as above: https://github.com/ros-visualization/qt_gui_core/issues/105

Do you have to have `shiboken` installed?

**Comment by kjcole** [comment]
I'll start with the last question: Despite rumors to the contrary, not everything has moved to Python3, and for work, I have something that depends on Python2 and PySide, which, in turn appears to depend on a lot of shiboken stuff -- or is that "shi-broken"? ;-) I may just need to resort to doing this on a separate beastie or moving the PySide stuff off of the one where I'm doing ROS.

Apparently, my Google-fu was not as great as yours: I may not have pared down the search terms enough, but when I searched I did not find anything I could identify as relevant. So, thanks for that.

Working backwards... during the hours-long compile, I noticed quite a few times that environment variables PYTHON_this and PYTHON_that and PYTHON_the_other_thing were being ignored / unused a lot, but I'll now attempt with the modifications to PKGBUILD that you suggest... Stay tuned. ;-)

**Comment by jwhendy** [comment]
> I'll start with the last question: Despite rumors to the contrary, not everything has moved to Python3, and for work, I have something that depends on Python2 and PySide

Understood, though *this* repo/ecosystem only relies on/uses `python3`. That said, arch installs are sometimes mixed packages and perhaps `shiboken` is pulling in both py2/py3 variants? I left it out of my commentary, but there was another issue about needing to accommodate `pyside2` because on systems where it's installed, it'll get pulled in and there was a breaking update. Perhaps the same for `shiboken`.

Re. searching, I just pasted in that exact error message and those are what popped up. And actually, I was too quick. there was really just [this](https://github.com/ros-visualization/qt_gui_core/issues/105) which linked to the ROS answers and another github link. It's all good.

Re. the python vars, indeed. A script added all those options in, but most don't use them. Since they showed up here, it was just a first guess. That said, from the links above, I think this will not be so easy a fix and probably requires an upstream report.

Do you have a working `catkin` workspace + `ros-melodic-ros-base`? I might suggest cloning upstream sources and building via `catkin build` or similar and then posting the issue there. If it's really a ROS issue with `shiboken`, we'll need upstream to fix anyway (and they should know about it and do so).

Let me know what you find!

**Comment by kjcole** [comment]
I had not heard of `catkin` until I tried installing `ros-melodic-desktop-full` and it pulled it in. I appear to have that at least:

```
$ catkin_<TAB><TAB>
catkin_create_pkg  catkin_generate_changelog  catkin_prepare_release  catkin_test_changelog
catkin_find_pkg    catkin_package_version     catkin_tag_changelog       

$ pacman -Q | grep catkin
python-catkin_pkg 0.4.14-1
ros-melodic-catkin 0.7.19-1
```

However, starting from `ros-melodic-desktop-full` without any prior ROS packages, it never got to `ros-melodic-ros-base`. 

But I just installed that now, after you inquired, and it appeared to install without issue.

**Comment by jwhendy** [comment]
Ah, you started with `ros-melodic-desktop-full`. Gotcha. I guess that will just pull everything in independently. I built meta packages from the ground up, which could be an alternative for you so you get to nice milestones vs. failing on a monster like `desktop-full`: `ros-melodic-ros-base`, `ros-melodic-desktop`, `ros-melodic-desktop-full` and (if you need) `ros-melodic-moveit`.

Looks like nothing installed `catkin_make` for you yet, but I would recommend `python-catkin_tools`. See https://robotics.stackexchange.com/questions/16604/ros-catkin-make-vs-catkin-build.

**Comment by acxz** [comment]
Hmm, apparently the `ros-melodic-qt-*` packages we have on the AUR are outdated. Maybe an updated version fixes it? Although based on the changelog I doubt it.

So I can install the package perfectly fine with `shiboken2` and `python-shiboken2`. I would move your projects away from `shiboken`, `python-shiboken`, and `python2-shiboken` if you can, time to move on.
In fact based on this comment: https://github.com/ros-visualization/qt_gui_core/issues/105#issuecomment-324391584 QT4 is not supported by ros-melodic. All three packages mentioned above require QT4. This means that upstream themselves are not supporting the setup you have on your machine. I'm afraid you are on your own for this. (Sure, just because they don't support it doesn't mean they have to let the build to break if QT4 based shiboken is on the system, but thats what it is.) Again I recommend to move your project over to the newer versions of `shiboken` and `python-shiboken` I'll wait for @jwhendy to respond, but I'm leaning on putting the wontfix label on this one.

**Comment by kjcole** [comment]
@acxz C'est la vie. ;-) Go ahead and label as you see fit. If you see a fix, great. But if not, well, I've been given some instruction on how to proceed.

**Comment by acxz** [comment]
So I thought I might as well try to reproduce it if I could:

`yay -Ss shiboken`:
```
aur/python-shiboken2-git r6554.6eb583d7-1 (+2 0.38%) 
    Generates bindings for C++ libraries using CPython source code (git version)
aur/shiboken2-git r6554.6eb583d7-1 (+2 0.38%) 
    Generates bindings for C++ libraries using CPython source code (git version)
aur/shiboken 1.2.4-7 (+8 0.36%) (Installed)
    CPython bindings generator for C++ libraries
aur/python-shiboken 1.2.4-7 (+8 0.36%) (Installed)
    Support library for Python bindings
aur/python2-shiboken 1.2.4-7 (+8 0.36%) (Installed)
    Support library for Python2 bindings
extra/shiboken2 5.14.0-3 (678.2 KiB 1.9 MiB) (Installed)
    Generates bindings for C++ libraries using CPython source code
extra/python-shiboken2 5.14.0-3 (163.6 KiB 404.4 KiB) (Installed)
    Generates bindings for C++ libraries using CPython source code
```
And then with a clean ROS build, i.e. no ROS packages installed on my system, I was able to successfully install `ros-melodic-qt-gui-cpp`. So I guess it does work.

TLDR: I cannot reproduce the error.

**Comment by kjcole** [comment]
@acxz I see the same thing from `yay -Ss shiboken`.  So, I'll look elsewhere for the trouble. (Maybe some `pip` command screwed around with what `pacman` / `yay` installed...) Meanwhile, label as you see fit.

**Comment by acxz** [comment]
Is your system completely update? Have you tried after a `yay -Syu`?
I'll just remove the build error and add the question label, since it seems like a problem only on your system. Also yes make sure nothing fishy is happening with pip.

**Comment by acxz** [comment]
Ahh I got the reproduction: It seems the culprit is `pyside2` and `pyside2-tools` not shiboken.
```
[ 52%] Running SIP generator for qt_gui_cpp_sip Python bindings...
sip: Deprecation warning: qt_gui_cpp.sip:1: %Module version number should be specified using the 'version' argument
qt.shiboken: (typesystem) Error: /home/acxz/.cache/yay/ros-melodic-qt-gui-cpp/src/qt_gui_core-0.3.14/qt_gui_cpp/src/qt_gui_cpp_shiboken/typesystem.xml:7:59: Nesting types by specifying '::' is no longer supported (qt_gui_cpp::CompositePluginProvider).
Cannot parse file: typesystem.xmlshiboken: Error running ApiExtractor.
Command line: --generatorSet=shiboken --include-paths=/home/acxz/.cache/yay/ros-melodic-qt-gui-cpp/src/qt_gui_core-0.3.14/qt_gui_cpp/src/qt_gui_cpp_shiboken/../../include:/usr/include/qt/:/usr/include/qt/QtWidgets:/usr/include/qt/QtGui:/usr/include/qt/QtCore:/usr/lib/qt//mkspecs/linux-g++ --typesystem-paths=/usr/share/PySide2/typesystems --output-directory=/home/acxz/.cache/yay/ros-melodic-qt-gui-cpp/src/build/src/qt_gui_cpp_shiboken global.h typesystem.xml
make[2]: *** [src/qt_gui_cpp_shiboken/CMakeFiles/qt_gui_cpp_shiboken.dir/build.make:75: src/qt_gui_cpp_shiboken/libqt_gui_cpp_shiboken/libqt_gui_cpp_shiboken_module_wrapper.cpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:348: src/qt_gui_cpp_shiboken/CMakeFiles/qt_gui_cpp_shiboken.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 55%] Compiling generated code for qt_gui_cpp_sip Python bindings...
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[ 58%] Meta target for qt_gui_cpp_sip Python bindings...
[ 58%] Built target libqt_gui_cpp_sip
make: *** [Makefile:130: all] Error 2
```
Also seems like a duplicate to https://github.com/ros-melodic-arch/ros-melodic-qt-gui-cpp/issues/11. Gonna close that one in favor of keeping this one open.

**Comment by acxz** [comment]
It seems like to reproduce this error you need both `shiboken` stuff and `pyside2` stuff.
And we are back here: https://github.com/ros-visualization/qt_gui_core/issues/142#issuecomment-445409573, getting the same error as @jwhendy did a long time ago.

@kjcole do you need `pyside2` and `pyside2-tools` if not then the easiest fix is to uninstall those two. @jwhendy maybe we should add `pyside2` and `pyside2-tools` to the conflicts field, so that this issue doesn't affect users. (Of course leaving this issue open so we don't forget about it)

**Comment by kjcole** [comment]
@acxz removing it breaks `freecad` but there's an AppImage version of `freecad` that I might install in it's place. For now, I've removed all three (`pyside2`, `pyside2-tools` and `freecad`). That got me over the hump, though there was the warning:

```
CMake Warning at /opt/ros/melodic/share/python_qt_binding/cmake/shiboken_helper.cmake:23 (find_package):
  By not providing "FindPySide2.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "PySide2", but
  CMake did not find one.

  Could not find a package configuration file provided by "PySide2" with any
  of the following names:

    PySide2Config.cmake
    pyside2-config.cmake

  Add the installation prefix of "PySide2" to CMAKE_PREFIX_PATH or set
  "PySide2_DIR" to a directory containing one of the above files.  If
  "PySide2" provides a separate development package or SDK, be sure it has
  been installed.
```

**Comment by acxz** [comment]
Don't worry about the warning. I am sad to hear that it breaks freecad tho. I think we have to leave it to the upstream issue I mentioned in the comment right above to resolve it. I'm afraid there is not much we can do on our end at the moment.

EDIT: You can install pyside2 and pyside2-tools after install this package for freecad. Just when building this package you cannot have pyside2 on your system, but after it is build it should be fine.

**Comment by stertingen** [comment]
As long as this is broken, I'd suggest to disable the generation of PySide bindings by removing this line:
https://github.com/ros-visualization/qt_gui_core/blob/kinetic-devel/qt_gui_cpp/src/CMakeLists.txt#L5

During my spare time, I'll try to fix this, but it may take some time to test, so I cannot provide a fix quickly.

**Comment by acxz** [comment]
Resolved in #14 feel free to comment if the issue still shows up.
